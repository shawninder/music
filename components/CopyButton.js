import React, { Component } from 'react'
import PropTypes from 'prop-types'
import defaultProps from '../helpers/defaultProps'
import propTypes from '../helpers/propTypes'

import Clipboard from 'clipboard'

import colors from '../styles/colors'
import durations from '../styles/durations'
import tfns from '../styles/timing-functions'

class CopyButton extends Component {
  componentDidMount () {
    this.clipboard = new Clipboard(this.el, {
      target: () => {
        return this.copyLink
      }
    })
    this.clipboard.on('success', (event) => {
      // event.clearSelection()
      this.props.notify({
        id: `party.linkCopied${Math.random()}`,
        body: this.props.dict.get('party.linkCopied'),
        duration: 2000
      })
    })
    this.clipboard.on('error', (event) => {
      console.error('Clipboard error', event)
      // TODO
    })
  }
  componentWillUnmount () {
    this.clipboard.destroy()
  }
  render () {
    const classes = this.props.className ? this.props.className.split(' ') : []
    classes.push('copyButton')
    return (
      <React.Fragment>
        <span
          className='copyLink'
          key='copyLink'
          ref={(el) => {
            this.copyLink = el
          }}
        >
          {global.location && global.location.href}
        </span>
        <a
          className={classes.join(' ')}
          key='copyButton'
          ref={(el) => {
            this.el = el
          }}
        >
          {this.props.dict.get('party.copyBtn')}
        </a>
        <style jsx>{`
          .copyLink {
            opacity: 0;
            position: absolute;
            z-index: -1;
          }
          .copyButton {
            background: transparent;
            border: 0;
            cursor: pointer;
            text-decoration: underline;
            padding: 3px 5px;
            opacity: 0;
            transition-property: opacity;
            transition-duration: ${durations.instant};
            transition-timing-function: ${tfns.easeInOutQuad};
            color: ${colors.text};
          }
          .copyButton.enabled {
            opacity: 1;
          }
        `}</style>
      </React.Fragment>
    )
  }
}

const props = [
  { name: 'dict', type: PropTypes.object.isRequired },
  { name: 'className', type: PropTypes.string, val: '' },
  { name: 'notify', type: PropTypes.func, val: console.log }
]

CopyButton.defaultProps = defaultProps(props)
CopyButton.propTypes = propTypes(props)

export default CopyButton
