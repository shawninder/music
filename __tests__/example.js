import { shallow } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'

import CopyButton from '../components/CopyButton'

describe('With Enzyme', () => {
  it('CopyButton shows "copy" by default', () => {
    const app = shallow(<CopyButton />)

    expect(app.find('.copyButton').text()).toEqual('copy')
  })
})

describe('With Snapshot Testing', () => {
  it('CopyButton shows "copy" by default', () => {
    const component = renderer.create(<CopyButton />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
