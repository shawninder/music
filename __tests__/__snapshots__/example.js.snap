// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Snapshot Testing CopyButton shows "copy" by default 1`] = `
Array [
  <span
    className="jsx-4241117872 copyLink"
  >
    
  </span>,
  <a
    className="jsx-4241117872 copyButton"
  >
    copy
  </a>,
]
`;
